### PR TITLE
unif: fix handling of eta

### LINKED ISSUE #106: `syntax error in as` ?
The following gives me a very strange error (in elpi 1.13.7):

```prolog
type bar any -> any.

foo (bar (uvar as K)) :- print "hello".

main :- (foo (x \ bar u)).
```
outputs
```Fatal error: syntax error in as```

whereas almost any other input:
```prolog
main :- (foo (x \ bar2 u)).
main :- (foo Y).
main :- (foo (bar u)).
main :- (foo (bar X)).
```
succeeds or fails without error.

Is this expected behavior?

### PR DESCRIPTION
**Pull request by gares**
This PR fixed the handling of eta when the term in eta long form is flexible.

closes #106

### CODE REVIEW

**Comment by nguermond** [comment]
This seems to break the following

```prolog
foo (bar (uvar K Args)) :- print K "@" Args.
main :- (foo (bar (x \ u))).
```
with error
```
Fatal error: as clause too complex
```
in `fix-106`, whereas it failed successfully in `master`.

**Comment by gares** [comment]
should be fixed now

**Comment by nguermond** [comment]
Sorry that I keep finding problems, but this now breaks
```prolog

pred patternify i:A, o:A.

patternify (uvar K Args) U :- print K "@" Args, (distinct_names Args).

%% master: success
%% fix-106: success
main :- patternify X U.

%% master: fails
%% fix-106: Fatal error: (Data.Term.Lam Data.Term.Nil) is not a list
main :- patternify (x \ X) U.

%% master: fails
%% fix-106: Fatal error: (Data.Term.Lam (Data.Term.Cons ((Data.Term.Const c0), Data.Term.Nil))) is not a list
main :- patternify (x \ X x) U.

%% master: fails
%% fix-106: Fatal error: (Data.Term.Lam
%%   (Data.Term.Cons ((Data.Term.Const u),
%%      (Data.Term.Cons ((Data.Term.Const c0), Data.Term.Nil))))) is not a list
main :- patternify (x \ X u x) U.
```

**Comment by gares** [comment]
Thanks for the tests, I will eventually get it right ;-)
I'll try again tomorrow

**Comment by gares** [comment]
This seems a bug in master, no?
```
%% master: fails
%% fix-106: Fatal error: (Data.Term.Lam (Data.Term.Cons ((Data.Term.Const c0), Data.Term.Nil))) is not a list
main :- patternify (x \ X x) U.
```

**Comment by nguermond** [comment]
Yes, I thought so too...

**Comment by gares** [comment]
one more try

**Comment by nguermond** [comment]
One more little snag:
```prolog
bar (uvar as K) :- print K.
bar2 (uvar K) :- print K.

%% master : fails
%% fix-106 : success ??
main :- bar (x \ X).

%% master : fails
%% fix-106 : success ??
main :- bar2 (x \ X).
```

**Comment by gares** [comment]
these are thought, thanks for posting them. I think I'll need yo rethink the code, since here the failures come from the fact that x is not in scope... maybe I should not eta expand but rater eta contract, ir at least fail if I cant eta contract...

**Comment by gares** [comment]
here it is, please test

**Comment by nguermond** [comment]
The following seems to hang:
```prolog
pred foo i:any.
pred foo2 i:any.

foo (uvar K Args).
foo2 (uvar K).

%% main: fails
%% fix-106: does not halt??
main :- foo (x \ y \ X x y).
main :- foo2 (x \ y \ X x y).

%% main: fails
%% fix-106: fails
main :- foo (x y \ X x y).
main :- foo2 (x y \ X x y).
```
Should these succeed?
Otherwise this seems to work well for all practical purposes.

**Comment by gares** [comment]
Here the terms are really `x` applied to `y\ X x y`. While `pi` and `sigma` support multiple binders at once, lambda alone can't
```
main :- foo (x y \ X x y).
main :- foo2 (x y \ X x y).
```

**Comment by gares** [comment]
The code I wrote sees `(x \ y \ X x y)` as the eta expansion of `X` so it passes. Do you find this surprising?

**Comment by nguermond** [comment]
What I mean is that `foo (x \ y \ X x y)` does not terminate for me.

**Comment by gares** [comment]
Right, sorry it was a very silly typo

**Comment by nguermond** [comment]
It seems `X` is substituted for `c \ X1` rather than `c \ X1 c`:
```prolog
pred foo i:any.

foo X :- print X, fail.
foo (uvar as X) :- print X.

main :- (foo (x \ X x)),
		print X.
```
outputs
```
c0 \ X0 c0
X1
c0 \ X1
```

**Comment by gares** [comment]
Many thanks for your help!
I did rewrite the whole code. It is more correct I believe and passes these tests, but there is also a bug I won't be able to fix. In particular occur-check is not up to eta, that is `(x\ X x) = X` fails.

The `as` clause has changed a little (w.r.t. the previous attempts) so that if you name a term like `x\X x`, you get that term, even if you write `uvar as T`. I think this is better, since `as` and `eta` are not orthogonal. Then the `var` builtin was also fixed to eta-contract what you give, so a rule like `foo (uvar as T) :- var T, print T.` works on `foo (x\ X x)` but prints `c0\ X c0` unlike in previous commits.

**Comment by nguermond** [comment]
This doesn't seem so bad compared to what we get!

There still seems to be a little problem with
`main :- pi x \ (var (y \ X x y)).` failing whereas both
```
main :- pi x \ (var (y \ X c y)). 
main :- pi x z \ (var (y \ X x y)).
```
succeed (but this works as expected with `uvar`/`as`).

Another minor remark: should `declare_constraint` also accept eta-expanded variables?

**Comment by gares** [comment]
Thanks again for your tests. I did fix this glitch and also called eta_contract_flex in declare_constraint

File: CHANGES.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+# UNRELEASED
+
+- Runtime/FFI:
+  - fix handling of eta expanded unification variables. Many thanks to
+    @nguermond for testing this tricky case.
```
Comment (gares) [inline CHANGES.md:5]: @nguermond I can put your name if you prefer.

File: src/runtime.ml [hunk 2]
```diff
@@ -1467,6 +1469,97 @@ let error_msg_hard_unif a b =
...
+    | AppArg(i,args) when e.(i) != C.dummy ->
+      aux d (deref_appuv ~from:adepth ~to_:d args e.(i))
+    | Nil | CData _ | Discard | Arg _ | AppArg _ -> false
+  and auxs d = function
+    | [] -> false
+    | t :: ts -> aux d t || auxs d ts
+  in
+  x < d && aux d t
+
+let rec exta_contract_args ~orig_depth ~depth r args eat adepth e =
```
Comment (nguermond) [inline src/runtime.ml:1505]: is exta a typo?

**Comment by nguermond** [comment]
This seems good to me, thank you so much for all your work!

File: CHANGES.md (hunk 1 above)
Comment (nguermond) [reply CHANGES.md:5]: It's Nathan Guermond, thanks!

File: src/runtime.ml (hunk 2 above)
Comment (gares) [reply src/runtime.ml:1505]: thanks

**Pull request merged**